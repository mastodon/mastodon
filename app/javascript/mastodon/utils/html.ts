// NB: This function can still return unsafe HTML
export const unescapeHTML = (html: string) => {
  const wrapper = document.createElement('div');
  wrapper.innerHTML = html
    .replace(/<br\s*\/?>/g, '\n')
    .replace(/<\/p><p>/g, '\n\n')
    .replace(/<[^>]*>/g, '');
  return wrapper.textContent;
};
