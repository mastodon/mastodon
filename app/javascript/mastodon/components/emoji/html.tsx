import { useMemo } from 'react';

import classNames from 'classnames';

import type { CustomEmojiMapArg } from '@/mastodon/features/emoji/types';
import { isModernEmojiEnabled } from '@/mastodon/utils/environment';
import type {
  OnAttributeHandler,
  OnElementHandler,
} from '@/mastodon/utils/html';
import { htmlStringToComponents } from '@/mastodon/utils/html';
import { polymorphicForwardRef } from '@/types/polymorphic';

import { AnimateEmojiProvider, CustomEmojiProvider } from './context';
import { textToEmojis } from './index';

export interface EmojiHTMLProps {
  htmlString: string;
  extraEmojis?: CustomEmojiMapArg;
  className?: string;
  onElement?: OnElementHandler;
  onAttribute?: OnAttributeHandler;
}

export const ModernEmojiHTML = polymorphicForwardRef<'div', EmojiHTMLProps>(
  (
    {
      extraEmojis,
      htmlString,
      as: asProp = 'div', // Rename for syntax highlighting
      className = '',
      onElement,
      onAttribute,
      ...props
    },
    ref,
  ) => {
    const contents = useMemo(
      () =>
        htmlStringToComponents(htmlString, {
          onText: textToEmojis,
          onElement,
          onAttribute,
        }),
      [htmlString, onAttribute, onElement],
    );

    return (
      <CustomEmojiProvider emojis={extraEmojis}>
        <AnimateEmojiProvider
          {...props}
          as={asProp}
          className={className}
          ref={ref}
        >
          {contents}
        </AnimateEmojiProvider>
      </CustomEmojiProvider>
    );
  },
);
ModernEmojiHTML.displayName = 'ModernEmojiHTML';

export const LegacyEmojiHTML = polymorphicForwardRef<'div', EmojiHTMLProps>(
  (props, ref) => {
    const {
      as: asElement,
      htmlString,
      extraEmojis,
      className,
      onElement,
      onAttribute,
      ...rest
    } = props;
    const Wrapper = asElement ?? 'div';
    return (
      <Wrapper
        {...rest}
        ref={ref}
        dangerouslySetInnerHTML={{ __html: htmlString }}
        className={classNames(className, 'animate-parent')}
      />
    );
  },
);
LegacyEmojiHTML.displayName = 'LegacyEmojiHTML';

export const EmojiHTML = isModernEmojiEnabled()
  ? ModernEmojiHTML
  : LegacyEmojiHTML;
