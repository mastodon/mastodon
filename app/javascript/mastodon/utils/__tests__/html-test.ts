import * as html from '../html';

describe('html', () => {
  describe('unescapeHTML', () => {
    it('returns unescaped HTML', () => {
      const output = html.unescapeHTML(
        '<p>lorem</p><p>ipsum</p><br>&lt;br&gt;',
      );
      expect(output).toEqual('lorem\n\nipsum\n<br>');
    });
  });
});
