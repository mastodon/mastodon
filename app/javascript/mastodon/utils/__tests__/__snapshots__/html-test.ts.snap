// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html > htmlStringToComponents > copies attributes to props 1`] = `
[
  <a
    href="https://example.com"
    rel="nofollow"
    target="_blank"
  >
    link
  </a>,
]
`;

exports[`html > htmlStringToComponents > handles nested elements 1`] = `
[
  <p>
    lorem 
    <strong>
      ipsum
    </strong>
  </p>,
]
`;

exports[`html > htmlStringToComponents > ignores empty text nodes 1`] = `
[
  <p>
    <span>
      lorem     ipsum
    </span>
  </p>,
]
`;

exports[`html > htmlStringToComponents > respects allowedTags option 1`] = `
[
  <p>
    lorem 
    <em>
      dolor
    </em>
  </p>,
]
`;

exports[`html > htmlStringToComponents > respects maxDepth option 1`] = `
[
  <p>
    <span />
  </p>,
]
`;

exports[`html > htmlStringToComponents > returns converted nodes from string 1`] = `
[
  <p>
    lorem ipsum
  </p>,
]
`;

exports[`html > htmlStringToComponents > uses default parsing if onElement returns undefined 1`] = `
[
  <p>
    lorem ipsum
  </p>,
]
`;
